detectors/node/opentelemetry-resource-detector-instana/src/detectors/InstanaAgentDetector.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Detector,
+  Resource,
+  ResourceDetectionConfig,
+} from '@opentelemetry/resources';
+import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
+import * as http from 'http';
+
+class InstanaAgentDetector implements Detector {
+  readonly INSTANA_AGENT_DEFAULT_HOST = 'localhost';
+  readonly INSTANA_AGENT_DEFAULT_PORT = 42699;
+
+  async detect(_config?: ResourceDetectionConfig): Promise<Resource> {
+    const host =
+      process.env.INSTANA_AGENT_HOST || this.INSTANA_AGENT_DEFAULT_HOST;
+    const port = Number(
+      process.env.INSTANA_AGENT_PORT || this.INSTANA_AGENT_DEFAULT_PORT
+    );
+
+    const data = await this._retryHandler(host, port, 0);
+
+    return new Resource({
+      [SemanticResourceAttributes.PROCESS_PID]: data.pid,
+      'instana.host.id': data.agentUuid,
+      // NOTE: The agent reads this property to correlate the process with the service
+      'service.instance.id': data.pid.toString(),
+    });
+  }
+
+  private timeout(ms: number) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  private async _retryHandler(
+    host: string,
+    port: number,
+    tries: number
+  ): Promise<any> {
+    const MAX_TRIES = 3;
+
+    try {
+      return await this._fetchAgentData(host, port);
+    } catch (err) {
+      if (tries < MAX_TRIES) {
+        console.log(
+          `Retrying to connect to the Instana agent on ${host}:${port}....`
+        );
+
+        const retryTimeout = process.env.INSTANA_RETRY_TIMEOUT
+          ? Number(process.env.INSTANA_RETRY_TIMEOUT)
+          : 1000;
+
+        await this.timeout(retryTimeout);
+        return await this._retryHandler(host, port, tries + 1);
+      }
+
+      throw err;
+    }
+  }
+
+  private async _fetchAgentData(host: string, port: number): Promise<any> {
+    const agentTimeout = process.env.INSTANA_AGENT_TIMEOUT
+      ? Number(process.env.INSTANA_AGENT_TIMEOUT)
+      : 3 * 1000;
+
+    const payload = {
+      pid: process.pid,
+    };
+
+    const payloadStr = JSON.stringify(payload);
+    const contentLength = Buffer.from(payloadStr, 'utf8').length;
+
+    return new Promise((resolve, reject) => {
+      // agent needs more than 1s!
+      const timeoutId = setTimeout(() => {
+        req.destroy();
+        reject(new Error('Instana Agent request timed out.'));
+      }, agentTimeout);
```

Comment:
can we use the `timeout` option on the `http.request` call instead of implementing it?
https://nodejs.org/api/http.html#httprequestoptions-callback